File: Rust/src/solution/maximum_units.rs
use super::Solution;

impl Solution {
    pub fn maximum_units(box_types: Vec<Vec<i32>>, truck_size: i32) -> i32 {
	let mut box_types = box_types;
	box_types.sort_by(|a, b| b[1].cmp(&a[1])); // Sort by units per box in descending order

	let mut total_units = 0;
	let mut remaining_truck_size = truck_size;

	for box_type in box_types {
	    let num_boxes = box_type[0];
	    let units_per_box = box_type[1];

	    if remaining_truck_size == 0 {
		break; // Truck is full
	    }

	    let boxes_to_take = num_boxes.min(remaining_truck_size);
	    total_units += boxes_to_take * units_per_box;
	    remaining_truck_size -= boxes_to_take;
	}

	total_units
    }
}

File: Dart/lib/maximumUnits.dart
import 'dart:math';

class Solution {
  int maximumUnits(List<List<int>> boxTypes, int truckSize) {
    int totalUnits = 0;
    boxTypes.sort((a, b) => b[1].compareTo(a[1]));
    for (List<int> boxType in boxTypes) {
      int numberOfBoxes = boxType[0];
      int unitsPerBox = boxType[1];
      int boxesToLoad = min(truckSize, numberOfBoxes);
      totalUnits += boxesToLoad * unitsPerBox;
      truckSize -= boxesToLoad;
      if (truckSize == 0) {
        break;
      }
    }
    return totalUnits;
  }
}
